src/requests.rs

```diff
@@ -0,0 +1,339 @@
+    engine::{self, general_purpose},
+    Engine as _,
+};
+use log::info;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::hash::Hash;
+use std::ops::Range;
+
+const ENDPOINT: &str = "https://pb.mfloto.com/v1/";
+
+/// Represents a temporary user returned by the login API.
+#[derive(Deserialize, Debug)]
+struct TempUser {
+    id: u32,
+    name: String,
+    role: u64,
+}
+
+/// Represents a temporary user used to create a new user.
+#[derive(Deserialize, Debug, Serialize, Clone)]
+pub struct TempCreationUser {
+    pub(crate) name: String,
+    pub(crate) password: String,
+    pub(crate) role: u64,
+}
+/// Implements the `default` trait for `TempCreationUser`.
+impl Default for TempCreationUser {
+    fn default() -> Self {
+        TempCreationUser {
+            name: String::new(),
+            password: String::new(),
+            role: PlantBuddyRole::NotLoggedIn.into(),
+        }
+    }
+}
+
+/// Represents the result of a request.
+pub type RequestResult<T> = Result<T, String>;
+
+/// Logs in a user with the given username and password.
+///
+/// # Arguments
+///
+/// * `username` - A string slice that holds the username.
+/// * `password` - A string slice that holds the password.
+///
+/// # Returns
+///
+/// Returns a `TempCreationUser` struct representing the logged-in user.
+pub async fn login(username: String, password: String) -> RequestResult<TempCreationUser> {
+    info!("Login Server request");
+    let client = reqwest::Client::new();
+    let response = client
+        .get(ENDPOINT.to_string() + "user/login")
+        .header(
+            "Authorization",
+            "Basic ".to_string() + &encode_credentials(username.clone(), password.clone()),
+        )
+        .send()
+        .await
+        .map_err(|e| e.to_string())?;
+
+    let result = response.error_for_status_ref().map(|_| ());
+
+    match result {
+        Ok(_) => {
+            let res = response.text().await.map_err(|e| e.to_string())?;
+            let v: Value = serde_json::from_str(&res).unwrap();
+            let role_value = v["role"]
+                .as_u64()
+                .ok_or("Role not found or not an integer")
+                .unwrap();
+
+            let login_user = TempCreationUser {
+                name: username.clone(),
+                password: password.clone(),
+                role: role_value.clone(),
+            };
+            info!("Login successful");
+            Ok(login_user)
+        }
+        Err(e) => {
+            info!("Login failed");
+            Err(e.to_string())
+        }
+    }
+}
+
+/// Gets all users with the given username and password.
+///
+/// # Arguments
+///
+/// * `username` - A string slice that holds the username.
+/// * `password` - A string slice that holds the password.
+///
+/// # Returns
+///
+/// Returns a vector of `User` structs representing all the users.
+pub async fn get_all_users(username: String, password: String) -> RequestResult<Vec<User>> {
```

Comment:
Given locking could be a problem since the UI code isn't async a Client could be passed around to the sites and methods on it could be called instead of giving nearly every request the same arguments.
```rs
struct Client{
	username: String,
	password: String
}
impl Client{
	pub async fn login() -> RequestResult<Self>;
	pub async fn get_all_users(&self) -> RequestResult<UserList>;
```